lib/three_scale/middleware/multitenant.rb

```diff
@@ -75,14 +75,26 @@ def master?
 
           @master ||= ::Account.unscoped.master
 
+          @is_master = session_of_master? || provider_key_of_master? || token_of_master?
+        end
+
+        def session_of_master?
           # this is supposed to match how we get the user_session in app/lib/authenticated_system/request.rb
           # on API calls cookies are not present though, so we need to use safe navigation
-          @user_session ||= UserSession.authenticate(@env['action_dispatch.cookies']&.signed&.public_send(:[], :user_session))
+          user_session ||= UserSession.authenticate(@env['action_dispatch.cookies']&.signed&.public_send(:[], :user_session))
+          user_session&.user&.account == @master
+        end
+
+        def provider_key_of_master?
+          param_places = %w[action_dispatch.request.query_parameters action_dispatch.request.request_parameters]
+          possible_keys = %w[provider_key api_key]
+          param_places.product(possible_keys).any? { |params, key| @env[params][key] == @master.api_key }
```

Comment:
```suggestion
          param_places.product(possible_keys).any? { |params, key| @env[params][key] == @master.provider_key }
```

I would use `provider_key` as it's a more common term, and hence more understandable.

Reply:
I'm not sure it is correct to say `provider_key` in the context of master. But since the method name is like that, that's alright. Or maybe good to see that `#api_key` method exists... anyway, merging your suggestion , doesn't make much difference for me.